subsystem_controllers/src/guidance_controller/plugin_manager.cpp

```diff
@@ -0,0 +1,434 @@
+        {
+            res->newstate = false;
+            return;
+        }
+
+        auto result_state = plugin_lifecycle_mgr_->transition_node_to_state(lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE, requested_plugin->name_, service_timeout_, call_timeout_);
+
+        bool activated = (result_state == lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE);
+
+        Entry updated_entry(requested_plugin->available_, activated, requested_plugin->name_, requested_plugin->type_, requested_plugin->capability_, true); // Mark as user activated
+        em_.update_entry(updated_entry);
+
+        res->newstate = activated;
+    }
+
+    void PluginManager::update_plugin_status(carma_planning_msgs::msg::Plugin::UniquePtr msg)
+    {
+        RCLCPP_DEBUG_STREAM(rclcpp::get_logger("guidance_controller"), "received status from: " << msg->name);
+        boost::optional<Entry> requested_plugin = em_.get_entry_by_name(msg->name);
+
+        if (!requested_plugin) // This is a new plugin so we need to add it
+        {
+            Entry plugin(msg->available, msg->activated, msg->name, msg->type, msg->capability, false);
+            add_plugin(plugin);
+            return;
+        }
+
+        Entry plugin(msg->available, msg->activated, msg->name, msg->type, msg->capability, requested_plugin->user_requested_activation_);
+
+        em_.update_entry(plugin);
+    }
+
+    bool PluginManager::matching_capability(const std::vector<std::string>& base_capability_levels, const std::vector<std::string>& compared_capability_levels)
+    {
+        for (size_t i=0; i < base_capability_levels.size() &&  i < compared_capability_levels.size(); i++)
+        {
+            if (compared_capability_levels[i].compare(base_capability_levels[i]) != 0)
+                return false;
+        }
+
+        return true;
+    }
+
+    void PluginManager::get_control_plugins_by_capability(SrvHeader, carma_planning_msgs::srv::GetPluginApi::Request::SharedPtr req, carma_planning_msgs::srv::GetPluginApi::Response::SharedPtr res)
+    {
+        std::vector<std::string> req_capability_levels;
+        boost::split(req_capability_levels, req->capability, boost::is_any_of("/"));
+
+        for(const auto& plugin : em_.get_entries())
+        {
+
+            std::vector<std::string> plugin_capability_levels; 
+            boost::split(plugin_capability_levels, plugin.capability_, boost::is_any_of("/"));
+
+            if(plugin.type_ == carma_planning_msgs::msg::Plugin::CONTROL && 
+                (req->capability.size() == 0 || (matching_capability(plugin_capability_levels, req_capability_levels) == 0 && plugin.active_ && plugin.available_)))
+            {
+                RCLCPP_DEBUG_STREAM(rclcpp::get_logger("guidance_controller"), "discovered control plugin: " << plugin.name_);
+                res->plan_service.push_back(plugin.name_ + plugin.capability_);
```

Comment:
This might be something to consider. Currently the capabilities aren't directly linked to the service names like this. E.g. even if a plugin reports "plan_tactical/plan_trajectory/lane_change" the service that is used to invoke that capability is still the plugin's planning service named "/guidance/plugins/PLUGIN_NAME/plan_trajectory". In the existing implementation the value that's pushed into the list here is `plugin.name_ + control_service_suffix_` .

I think that helps with arbitrator and plan delegator having consistent interfaces by which to query plugins during planning, but I could see benefits to making it more granular. We'd have to adjust arbitrator and plan_delegator a bit. Do you think this is worth discussing in more detail?

Reply:
I think this was a bit of an oversight on my part. I'm fine with how we had it before. If we took the approach here, then it would be difficult for the receiver to disambiguate the plugin name from the service call which could make it unclear what the original capability was.